riscv/src/interrupt.rs

```diff
@@ -2,187 +2,67 @@
 
 // NOTE: Adapted from cortex-m/src/interrupt.rs
 
-pub mod machine {
-    use crate::register::{mepc, mstatus};
+use crate::result::Result;
 
-    /// Disables all interrupts in the current hart (machine mode).
-    #[inline]
-    pub fn disable() {
-        // SAFETY: It is safe to disable interrupts
-        unsafe { mstatus::clear_mie() }
-    }
+// re-export useful riscv-pac traits
+pub use riscv_pac::{CoreInterruptNumber, ExceptionNumber, InterruptNumber};
 
-    /// Enables all the interrupts in the current hart (machine mode).
-    ///
-    /// # Safety
-    ///
-    /// Do not call this function inside a critical section.
-    #[inline]
-    pub unsafe fn enable() {
-        mstatus::set_mie()
-    }
+pub mod machine;
+pub mod supervisor;
 
-    /// Execute closure `f` with interrupts disabled in the current hart (machine mode).
-    ///
-    /// This method does not synchronise multiple harts, so it is not suitable for
-    /// using as a critical section. See the `critical-section` crate for a cross-platform
-    /// way to enter a critical section which provides a `CriticalSection` token.
-    ///
-    /// This crate provides an implementation for `critical-section` suitable for single-hart systems,
-    /// based on disabling all interrupts. It can be enabled with the `critical-section-single-hart` feature.
-    #[inline]
-    pub fn free<F, R>(f: F) -> R
-    where
-        F: FnOnce() -> R,
-    {
-        let mstatus = mstatus::read();
+#[cfg(not(feature = "s-mode"))]
+pub use machine::*;
+#[cfg(feature = "s-mode")]
+pub use supervisor::*;
 
-        // disable interrupts
-        disable();
+/// Trap Cause
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Trap<I, E> {
+    Interrupt(I),
+    Exception(E),
+}
```

Comment:
In my opinion, this use of generics is much harder to use for not that much gain.

Is there a reason you decided to redefine `Trap` using generics for the interrupt and exception number?

If this is to allow external users to use a custom `Interrupt` and `Exception` type, maybe we could also provide a top-level alias that defaults to the local `Interrupt` and `Exception` type?

Reply:
Yes, the idea now is updating svd2rust so PACs do this kind of type binding in pac::interrupt. PACs could either re-export the Exception and CoreInterrupt enums that are in riscv or define their custom type. Also the interface with mcause can be adjusted to hide this complexity to final users.

I'll do a PoC and share it with you.